Misc clean up, necessary for more Yask integration

FabioLuporini opened the pull request:
Many Operator-related improvements, with some refactoring.

Also extends the hierarchy in ``interfaces.py``

FabioLuporini commented:
@navjotk here I centralised arguments code generation inside ``nodes.Function``, which to me makes much more sense than having code generation methods attached to ``Arguments``.
In the future we may think of creating an actual tree visitor to do code generation (low priority now)

navjotk commented on devito/arguments.py:
```diff
@@ -79,10 +74,6 @@ def __init__(self, name, provider, reducer, default_value=None):
-    def decl(self):
-        return c.Value('const %s' % c.dtype_to_ctype(self.dtype), self.name)
-
     def verify(self, value):
```
I don't understand why `decl` is being removed here.

navjotk commented on devito/arguments.py:
```diff
@@ -186,14 +177,6 @@ def rtargs(self):
-        else:
-            return self.rtargs[0].ccode
-
-    @property
```
Why is this being removed?

navjotk commented on devito/arguments.py:
```diff
@@ -102,23 +93,7 @@ class TensorArgument(Argument):
-                             '(%s (*)%s) %s' % (c.dtype_to_ctype(self.dtype),
-                                                shape, '%s_vec' % self.name))
-        return cast
+        super(TensorArgument, self).__init__(name, provider, provider)
```
Same. Why is this being removed?

navjotk commented on devito/dimension.py:
```diff
@@ -31,10 +33,10 @@ def __str__(self):
+        if self.size is not None:
+            return Number(self.size)
+        else:
+            return Symbol('%s%s' % (self.name, Dimension._size_suffix))
```
This seems to be undoing the work I did with arguments.

navjotk replied:
This should be using the `rtargs` framework.

navjotk commented on devito/nodes.py:
```diff
@@ -547,16 +549,51 @@ def __repr__(self):
+                argtypes.append(ctypes.c_void_p)
+        return argtypes
+
+    @property
```
Very ugly hacky method. Should be moved inside each argument

navjotk commented on devito/nodes.py:
```diff
@@ -547,16 +549,51 @@ def __repr__(self):
+                                     '*restrict %s_vec' % i.name))
+            else:
+                decls.append(c.Value('void', '*_%s' % i.name))
+        return decls
```
Same as above. Undoes the cleanup I did earlier and brings us back to this ugly hacky if-else mess.

navjotk commented on devito/nodes.py:
```diff
@@ -547,16 +549,51 @@ def __repr__(self):
+                lvalue = c.Pointer(c.Value(ctype, i.name))
+                rvalue = '(%s*) %s' % (ctype, '_%s' % i.name)
+                ccasts.append(c.Initializer(lvalue, rvalue))
+        return ccasts
```
Same as above. This is not a cleanup. Quite the contrary.

mlange05 commented:
@navjotk Ok, I think I can provide some context here from previous discussions. The problem with the current argument derivation hierarchy is that it breaks abstraction and encapsulation, because it contains a whole bunch of _compile-time_ properties (types, declarations, etc.), despite encapsulating _runtime_ behaviour. I believe this PR is trying to decouple the codegen-specific things from the runtime argument hierarchy, because these things will eventually need to be sub-classed for different compiler backends (eg. YASK). The intent here, I believe, is to make the argument derivation engine agnostic to the compiler backend, so as to avoid having to have YASK-specific argument types. As such, I think this is actually much cleaner, since it increases the overall degree of encapsulation and draws rigorous lines between our internal abstractions - which is ultimately the goal of object-orientation.

Admittedly, these discussions would benefit from clearer developer-level documentation of the internal hierarchies, but they are still very much in flux.

FabioLuporini commented:
@mlange05 is right about the rationale behind this.

Moving everything inside the node hierarchy is a first step towards building a tree visitor doing code generation. Once the tree visitor will be ready, we won't even have those ``if/then/else``.

BTW, we can argue about lots of things, like these edits making the structure more or less clean, but definitely this is not hacky.

FabioLuporini commented:
**UPDATE**

Before this PR, code generation was annoyingly distributed over multiple modules: ``operator.py``, the ``Argument`` classes in ``arguments.py``, the node classes in ``nodes.py``. Now, this has been fixed. There's a new tree visitor responsible for code generation. It basically aggregates the ``ccode``, ``cdecl``, ``ctop``, ... . I added a test for this visitor.

The second update is that I refactored the crazy messy ``Operator`` initialization, which fixes issue #318.

FabioLuporini commented:
@navjotk this needs another round of reviewing

navjotk commented on devito/operator.py:
```diff
@@ -165,17 +145,17 @@ def arguments(self, **kwargs):
+            if i.is_TensorArgument:
+                assert(i.verify(kwargs.pop(i.name, None)))
+            else:
+                i.verify(kwargs.pop(i.name, None))
```
This is changing the tree structure of the arguments derivation. Why is this required? Especially in this PR?

FabioLuporini replied:
Can you provide an example of what gets broken here? As all tests are passing, hence probably there's a flaw in our testing infrastructure.

navjotk replied:
The arguments derivation is a WIP, including tests. 

- You removed the tree structure by putting it all in one loop
- Scalar Arguments aren't being called any more
- Dimensions returned to the assumption that there is only one argument per dimension. Undoing my previous cleanup regarding this

FabioLuporini replied:
Again, I need an example, or **better**, a test to catch what's breaking here, otherwise the only option is to go ahead and "fix" this in a second moment, when the test will be available

My concerns follow:
* ``self.parameters`` includes both tensor and scalar aguments. So ``ScalarArgument``s are actually being called (``else`` branch).
* I don't see how I could have altered the tree structure, since you also had several loops
* I don't understand the point about dimensions, as it seems the same loop to me ?

I totally understand the arguments stuff was WIP; but still, this can't be my fault if no tests caught this "issue" (?)

FabioLuporini commented:
Addressed @navjotk 's comments after offline discussion

FabioLuporini commented:
Merging...

FabioLuporini commented:
Merged.

mloubout commented on devito/operator.py:
```diff
@@ -444,13 +406,21 @@ def _retrieve_stencils(self, expressions):
+        as well as all traversed dimensions.
+        """
+        input = flatten(terminals(i) for i in expressions)
+        input = filter_sorted([i.base.function for i in input], key=attrgetter('name'))
```
This breaks if there  is a `Dimension` in `terminals`
